octopusdeploy_framework/resource_deploymentfreeze.go

```diff
@@ -0,0 +1,213 @@
+	deploymentFreeze, err := mapFromState(ctx, plan)
+	if err != nil {
+		resp.Diagnostics.AddError("error while creating deployment freeze", err.Error())
+		return
+	}
+
+	createdFreeze, err := deploymentfreezes.Add(f.Config.Client, deploymentFreeze)
+	if err != nil {
+		resp.Diagnostics.AddError("error while creating deployment freeze", err.Error())
+		return
+	}
+
+	diags = mapToState(ctx, plan, createdFreeze, false)
+	if diags.HasError() {
+		return
+	}
+
+	diags = resp.State.Set(ctx, plan)
+	resp.Diagnostics.Append(diags...)
+}
+
+func (f *deploymentFreezeResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	var plan *deploymentFreezeModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	existingFreeze, err := deploymentfreezes.GetById(f.Config.Client, plan.ID.ValueString())
+	if err != nil {
+		if err := errors.ProcessApiErrorV2(ctx, &resp.State, plan, err, "deployment freeze"); err != nil {
+			resp.Diagnostics.AddError("unable to load deployment freeze", err.Error())
+		}
+		return
+	}
+
+	updatedFreeze, err := mapFromState(ctx, plan)
+	if err != nil {
+		resp.Diagnostics.AddError("error while mapping deployment freeze", err.Error())
+	}
+
+	updatedFreeze.SetID(existingFreeze.GetID())
+	updatedFreeze.Links = existingFreeze.Links
+
+	updatedFreeze, err = deploymentfreezes.Update(f.Config.Client, updatedFreeze)
+	if err != nil {
+		resp.Diagnostics.AddError("error while updating deployment freeze", err.Error())
+	}
+
+	diags := mapToState(ctx, plan, updatedFreeze, false)
+	if diags.HasError() {
+		return
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, plan)...)
+}
+
+func (f *deploymentFreezeResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	var state *deploymentFreezeModel
+	diags := req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	freeze, err := deploymentfreezes.GetById(f.Config.Client, state.GetID())
+	if err != nil {
+		if err := errors.ProcessApiErrorV2(ctx, &resp.State, state, err, "deployment freeze"); err != nil {
+			resp.Diagnostics.AddError("unable to load deployment freeze", err.Error())
+		}
+		return
+	}
+
+	err = deploymentfreezes.Delete(f.Config.Client, freeze)
+	if err != nil {
+		resp.Diagnostics.AddError("unable to delete deployment freeze", err.Error())
+	}
+
+	resp.State.RemoveResource(ctx)
+}
+
+func mapToState(ctx context.Context, state *deploymentFreezeModel, deploymentFreeze *deploymentfreezes.DeploymentFreeze, useSourceForDates bool) diag.Diagnostics {
+	state.ID = types.StringValue(deploymentFreeze.ID)
+	state.Name = types.StringValue(deploymentFreeze.Name)
+	if useSourceForDates {
+		state.Start = types.StringValue(deploymentFreeze.Start.Format(time.RFC3339))
+		state.End = types.StringValue(deploymentFreeze.End.Format(time.RFC3339))
+	}
```

Comment:
if the time contains a timezone the value for the times are returned with a server local timezone, for create and update operations we just store the value provided in the plan, for a refresh (read) operation it will store the value from the api as it may have been changed.